v0/destinations/snapchat_custom_audience/transform.js

```diff
@@ -0,0 +1,188 @@
+ */
+const validatePayload = message => {
+  if (!message.type) {
+    throw new ErrorBuilder()
+      .setMessage(
+        "[snapchat_custom_audience]::Message Type is not present. Aborting message."
+      )
+      .setStatus(400)
+      .build();
+  }
+  if (!message.properties) {
+    throw new ErrorBuilder()
+      .setMessage(
+        "[snapchat_custom_audience]::Message properties is not present. Aborting message."
+      )
+      .setStatus(400)
+      .build();
+  }
+  if (!message.properties.listData) {
+    throw new ErrorBuilder()
+      .setMessage(
+        "[snapchat_custom_audience]::listData is not present inside properties. Aborting message."
+      )
+      .setStatus(400)
+      .build();
+  }
+  if (message.type.toLowerCase() !== "audiencelist") {
+    throw new ErrorBuilder()
+      .setMessage(
+        `[snapchat_custom_audience]::Message Type ${message.type} not supported.`
+      )
+      .setStatus(400)
+      .build();
+  }
+  if (!message.properties.listData.add && !message.properties.listData.remove) {
+    throw new ErrorBuilder()
+      .setMessage(
+        "[snapchat_custom_audience]::Neither 'add' nor 'remove' property is present inside 'listData'. Aborting message."
+      )
+      .setStatus(400)
+      .build();
+  }
+};
+
+const responseBuilder = (metadata, message, { Config }, type) => {
+  const { schema, segmentId } = Config;
+  const payload = { users: [] };
+  const userPayload = { schema: [], data: [] };
+  let schemaType;
+  let hashedProperty;
+  // Normalizing and hashing ref: https://marketingapi.snapchat.com/docs/#normalizing-hashing
+
+  let userArray;
+  if (type === "add") {
+    userArray = message.properties.listData.add;
+  } else if (type === "remove") {
+    userArray = message.properties.listData.remove;
+  }
+
+  switch (schema) {
+    case "email":
+      schemaType = "EMAIL_SHA256";
+      userArray.forEach(element => {
+        hashedProperty = element?.email;
+        // what if email is not present in all cases?
+        hashedProperty = sha256(hashedProperty.toLowerCase().trim());
+        userPayload.data.push([hashedProperty]);
+      });
+      break;
+
+    case "phone":
+      schemaType = "PHONE_SHA256";
+      userArray.forEach(element => {
+        hashedProperty = element?.phone || element?.mobile;
+        hashedProperty = sha256(
+          hashedProperty
+            .toLowerCase()
+            .trim()
+            .replace(/^0+/, "")
+            .replace(/[^0-9]/g, "")
+        );
+        userPayload.data.push([hashedProperty]);
+      });
+      break;
+
+    case "mobileAdId":
+      schemaType = "MOBILE_AD_ID_SHA256";
+      userArray.forEach(element => {
+        hashedProperty =
+          element?.mobileId || element?.mobileAdId || element?.mobile_id;
+        hashedProperty = sha256(hashedProperty.toLowerCase());
+        userPayload.data.push([hashedProperty]);
+      });
+      break;
+
+    default:
+      throw new CustomError("Invalid schema", 400);
```

Comment:
Can we move the loop outside the switch case, otherwise we are repeating parts of code multiple times

Reply:
Here in these 3 cases, we are normalizing(hashing) each one differently.
Moving the loop outside the switch case will continuously iterate through the switch case(checking schema).

Reply:
Modified this part